pkg/api/helpers.go

```diff
@@ -494,15 +494,18 @@ func TolerationToleratesTaint(toleration *Toleration, taint *Taint) bool {
 }
 
 // TaintToleratedByTolerations checks if taint is tolerated by any of the tolerations.
-func TaintToleratedByTolerations(taint *Taint, tolerations []Toleration) bool {
-	tolerated := false
+func TaintToleratedByTolerations(taint *Taint, tolerations []Toleration) (bool, bool) {
```

Comment:
The kubeletAwareness thing that you're doing isn't quite right.

The semantics of this feature is "kubelet rejects the pod if the node has one or more taints of type NoScheduleNoAdmit that are not tolerated"

So the change you made here isn't correct. What you should do is rename the KubeletAwareness field in ErrTaintsTolerationsNotMatch to `SomeUntoleratedTaintIsNoAdmit`. Then you should have predicates.PodToleratesNodeTaints() set the value of that field when predicates.PodToleratesNodeTaints() returns false. The field should be set to true if the node has at least one NoScheduleNoAdmit taint that is not tolerated, and false otherwise. Then checkPredicateFitness() should reject the pod if predicates.PodToleratesNodeTaints() returns false and the SomeUntoleratedTaintIsNoAdmit field of the error is true.

Note that this means predicates.PodToleratesNodeTaints() cannot return false immediately when it finds the first un-tolerated taint. Instead, it can only return false once it finds the first un-tolerated NoScheduleNoAdmit taint (if one exists). If it returns false when it finds the first un-tolerated taint and that taint is of type NoSchedule, it does not know yet whether there is some later un-tolerated taint of type NoScheduleNoAdmit, so it cannot be sure how to set the SomeUntoleratedTaintIsNoAdmit field of the error. So the logic is like this (this is just pseudocode):
```
SomeUntoleratedTaintIsNoAdmit := false
fits := true
for (each taint T)
   if T is tolerated, continue
   // T is not tolerated
   fits := false
   if T is of type NoScheduleNoAdmit
      SomeUntoleratedTaintIsNoAdmit := true
      return